torch/distributed/_symmetric_memory/__init__.py

```diff
@@ -1493,4 +1488,192 @@ def rendezvous(
+    Supplying multiple :attr:`B` matrices allows for more computation to hide the
+    communication latency. When :attr:`B` is a list, :attr:`B_scale` must be a list of
+    the same size. Additionally, :attr:`bias`, :attr:`result_scale`, :attr:`out_dtype`
+    and :attr:`use_fast_accum` must either be ``None`` or a list of the same size as
+    :attr:`B`.
+
+    Arguments:
+        A_shard (Tensor): the shard to be all-gathered as the lhs tensor of the scaled
+            matmul.
+
+        B (Tensor or List[Tensor]): the rhs tensor(s) of the scaled matmul(s).
+
+        A_scale (Tensor): the lhs scale of the scaled matmul.
+
+        B_scale (Tensor or List[Tensor]): the rhs scale(s) of the scaled matmul(s).
+
+    Keyword args:
+        group (str | ProcessGroup): the process group with which ``A_shard`` is
+            all-gathered. This can be either a group name or a process group object.
+            Devices in the group must be connected via interconnects supported by
+            symmetric memory (e.g. NVLink).
+
+        gather_dim (int, optional): the dimension along which :attr:`A_shard` is
+            all-gathered.
+
+        bias (Tensor or List[Tensor | None], optional): the :attr:`bias` argument of
+            :func:`torch._scaled_mm`. Must be a tensor, ``None``, or a list of the same
+            size as :attr:`B`.
+
+        result_scale (Tensor or List[Tensor | None], optional): the :attr:`result_scale`
+            argument of :func:`torch._scaled_mm`. Must be a tensor, ``None``, or a list
+            of the same size as :attr:`B`.
+
+        out_dtype (torch.dtype or List[torch.dtype | None], optional): the
+            :attr:`out_dtype` argument of :func:`torch._scaled_mm`. Must be a
+            :class:`torch.dtype`, ``None``, or a list of the same size as :attr:`B`.
+
+        use_fast_accum (bool or List[bool]), optional): the :attr:`use_fast_accum`
+            argument of :func:`torch._scaled_mm`. Must be a ``bool``, ``None``, or a
+            list of the same size as :attr:`B`.
+
+    Returns:
+        A tuple containing the scaled matmul result(s) and the all-gather result. The
+        scaled matmul result(s) has the same type as ``B``.
+    """
+    # B must be a Tensor or a list of Tensor
+    if not isinstance(B, torch.Tensor) and not (
+        isinstance(B, Sequence) and all(isinstance(b, torch.Tensor) for b in B)
+    ):
+        raise TypeError("B must be a tensor or a list of tensors")
+
+    B_is_list = True
+    if isinstance(B, torch.Tensor):
+        B = [B]
+        B_is_list = False
+
+    def check_and_normalize_arg(
+        arg: Any, arg_name: str, arg_type: type, default: Any
+    ) -> Any:
+        if B_is_list:
+            if arg is None:
+                return [default] * len(B)
+            elif not isinstance(arg, Sequence):
+                raise ValueError(f"{arg_name} must be a list or None when B is a list")
```

Comment:
Micro-nit: we're not validating the individual types of the elements of the sequence.

Reply:
Note that doing so requires handling `[None]` correctly...